### PR TITLE
[KOGITO-4470] - ConfigMap required ownerReference as controller

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
Many thanks for submitting your Pull Request :heart:! 

Jira Issue : https://issues.redhat.com/browse/KOGITO-4470

As part of following PR, I had done below changes :
- Previously we are modifying user-provided configmap and append extra configuration in it. Now I had seperate them into two independent configmap. User-provided configmap mount to the deployment as it is without any change. To append extra configuration operator will create new config-map and mount to deployment.
- I took out configmap from deployer-resource -> createRequiredResources and create new indepenent reconfig conciler for configmap only.
- Previously we are creating protobuf configmap when kogitoruntime service start and append to data-index service later. Now we create & append protobuf configmap on startup of data-index service.

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've ran `make before-pr` and everything is working accordingly
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](https://github.com/kiegroup/kogito-operator/blob/master/RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

**Comment by sutaakar** [comment]
@vaibhavjainwiz I tried to deploy `process-quarkus-example` with persistence and events together with Data index. However (after several reconciliations once deployments stabilized) the protobuf files were added as volumes to KogitoRuntime Deployment. Data index Deployment didn't have any volume.

File: api/v1beta1/kogitoruntime_types.go [hunk 1]
```diff
@@ -119,7 +119,8 @@ type KogitoRuntimeList struct {
 func (k *KogitoRuntimeList) GetItems() []api.KogitoRuntimeInterface {
 	models := make([]api.KogitoRuntimeInterface, len(k.Items))
 	for i, v := range k.Items {
-		models[i] = &v
+		item := v
```
Comment (spolti) [inline api/v1beta1/kogitoruntime_types.go:122]: does it change something?

File: cmd/kogito/command/converter/config_converter.go [hunk 2]
```diff
@@ -30,6 +29,8 @@ const (
 	configMapSuffix     = "custom-properties"
 	createdByAnnonKey   = "createdBy"
 	createdByAnnonValue = "Kogito CLI"
+	// ConfigMapApplicationPropertyKey ...
```
Comment (spolti) [inline cmd/kogito/command/converter/config_converter.go:32]: not sure if ...Key fits the purpose of this var, maybe ConfigMapAppPropsFIleName?

File: cmd/kogito/command/deploy/deploy_service_test.go [hunk 3]
```diff
@@ -136,7 +136,7 @@ my.nice.property=socool
 	exists, err := kubernetes.ResourceC(ctx.GetClient()).Fetch(cm)
 	assert.NoError(t, err)
 	assert.True(t, exists)
-	assert.Contains(t, cm.Data[kogitoservice.ConfigMapApplicationPropertyKey], "quarkus.log.level")
+	assert.Contains(t, cm.Data[converter.ConfigMapApplicationPropertyKey], "quarkus.log.level")
```
Comment (spolti) [inline cmd/kogito/command/deploy/deploy_service_test.go:139]: maybe compare the full value here? property=value?

File: core/connector/protobuf.go [hunk 4]
```diff
@@ -18,146 +18,96 @@ import (
...
-	return nil
-}
 
-// MountProtoBufConfigMapOnDataIndex mounts protobuf configMaps from KogitoRuntime services into the given deployment instance of DataIndex
-func (p *protoBufHandler) MountProtoBufConfigMapOnDataIndex(kogitoRuntime api.KogitoRuntimeInterface) (err error) {
-	supportingServiceManager := manager.NewKogitoSupportingServiceManager(p.Context, p.supportingServiceHandler)
-	deployment, err := supportingServiceManager.FetchKogitoSupportingServiceDeployment(kogitoRuntime.GetNamespace(), api.DataIndex)
-	if err != nil || deployment == nil {
+	// check if data-index service not exists then return
+	if dataIndexService == nil {
```
Comment (spolti) [inline core/connector/protobuf.go:68]: maybe print some warning or debug message here?

File: core/connector/protobuf_configmap.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	if !available {
+		p.Log.Debug("deployment not available")
+		return nil
+	}
+
+	protobufEndpoint := p.kogitoServiceHandler.GetKogitoServiceEndpoint(runtimeInstance) + protobufSubdir
+	protobufListURL := protobufEndpoint + protobufListFileName
+	protobufListBytes, err := getHTTPFileBytes(protobufListURL)
+	if err != nil {
+		p.Log.Error(err, "failed to get protobuf file list")
```
Comment (spolti) [inline core/connector/protobuf_configmap.go:118]: append the file/url here or will it be printed on the err ?

File: core/infrastructure/configmap.go [hunk 6]
```diff
@@ -16,50 +16,190 @@ package infrastructure
...
+		deployment.Spec.Template.Spec.Containers[0].VolumeMounts =
+			append(deployment.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
+				Name:      cm.Name,
+				MountPath: mountPath,
+				SubPath:   fileName,
+			})
+	}
+}
+
+func isMountAsLiteral(configMap *corev1.ConfigMap) bool {
```
Comment (spolti) [inline core/infrastructure/configmap.go:187]: shouldn't it actually query the installed config maps ans return the ones with matches the label?
if, not probably the func name needs to be updated.

File: core/kogitoservice/app_configmap_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+		Context:             context,
+		instance:            instance,
+		configMapHandler:    infrastructure.NewConfigMapHandler(context),
+		deltaProcessor:      NewDeltaProcessor(context),
+		infraManager:        manager.NewKogitoInfraManager(context, infraHandler),
+		appConfigMapHandler: NewAppConfigMapHandler(context),
+	}
+}
+
+func (c *appConfigMapReconciler) Reconcile() (reconcileInterval time.Duration, err error) {
```
Comment (spolti) [inline core/kogitoservice/app_configmap_reconciler.go:57]: not sure if crateing a new reconcile for every config map kind is a good idea.
is it really necessary? can't we have just one reconciler to all config maps managed by the operator?

**Review by spolti — CHANGES_REQUESTED** [review]

File: api/v1beta1/kogitoruntime_types.go (hunk 1 above)
Comment (vaibhavjainwiz) [reply api/v1beta1/kogitoruntime_types.go:122]: when we directly assign `models[i] = &v` then pointer value of v is assign to `models[i]` so in the next iteration when value of v is change then value inside `models[i]` also override.
With `item` we had introduced local variable which create new for every iteration so value inside `models[i]` never change.

File: cmd/kogito/command/converter/config_converter.go (hunk 2 above)
Comment (vaibhavjainwiz) [reply cmd/kogito/command/converter/config_converter.go:32]: done

File: cmd/kogito/command/deploy/deploy_service_test.go (hunk 3 above)
Comment (vaibhavjainwiz) [reply cmd/kogito/command/deploy/deploy_service_test.go:139]: done

File: core/connector/protobuf_configmap.go (hunk 5 above)
Comment (vaibhavjainwiz) [reply core/connector/protobuf_configmap.go:118]: done

File: core/kogitoservice/app_configmap_reconciler.go (hunk 7 above)
Comment (vaibhavjainwiz) [reply core/kogitoservice/app_configmap_reconciler.go:57]: These are only internal reconciler which work as a service. Kogitoruntimr_controller receive reconcile request from olm and handed over to these service to perform its task.

File: core/connector/protobuf.go (hunk 4 above)
Comment (vaibhavjainwiz) [reply core/connector/protobuf.go:68]: done

**Comment by vaibhavjainwiz** [comment]
> @vaibhavjainwiz I tried to deploy `process-quarkus-example` with persistence and events together with Data index. However (after several reconciliations once deployments stabilized) the protobuf files were added as volumes to KogitoRuntime Deployment. Data index Deployment didn't have any volume. Those volumes should be attached to Data index Deployment.

@sutaakar I had fixed these issues. Could you please verify it again.

File: core/infrastructure/configmap.go (hunk 6 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/configmap.go:188]: @spolti I am always got confused while naming my function. This function reads configmap labels and check whether it created using literals or through file. 
Could you please suggest some better name.

**Comment by sutaakar** [comment]
Tested scenarios (will add more over the time):

1. Simple deployment without any CM:
```
apiVersion: app.kiegroup.org/v1beta1
kind: KogitoRuntime
metadata:
  name: process-quarkus-example
spec:
  image: 'quay.io/kiegroup/examples-process-quarkus-example-nightly:latest'
```
CM process-quarkus-example-properties is created, empty. :heavy_check_mark:
Modifying the generated CM:
```
data:
  some: thing
```
Modification is reflected in Kogito application - :heavy_multiplication_x:  - the modification is immediately overwritten

File: core/connector/protobuf_configmap_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	resources := make(map[reflect.Type][]resource.KubernetesResource)
+	labels := map[string]string{ConfigMapProtoBufEnabledLabelKey: "true"}
+	configMapList, err := p.configMapHandler.FetchConfigMapsForLabel(p.instance.GetNamespace(), labels)
+	if err != nil {
+		return nil, err
+	}
+	if len(configMapList.Items) > 0 {
+		kubernetesResources := make([]resource.KubernetesResource, len(configMapList.Items))
+		for i, configMap := range configMapList.Items {
+			kubernetesResources[i] = &configMap
```
Comment (sutaakar) [inline core/connector/protobuf_configmap_reconciler.go:101]:
```suggestion
			item := configMap
			kubernetesResources[i] = &item
```

File: core/infrastructure/configmap.go [hunk 9]
```diff
@@ -16,50 +16,205 @@ package infrastructure
...
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: configMap.GetName(),
+			},
+		},
+	}
+	deployment.Spec.Template.Spec.Containers[0].EnvFrom = append(deployment.Spec.Template.Spec.Containers[0].EnvFrom, envFromSource)
+}
+
+func (c *configMapHandler) RemoveConfigMapOwnership(key types.NamespacedName, owner resource.KubernetesResource) (err error) {
+	c.Log.Info("Removing Configmap ownership", "configmap", key.Name, "owner", owner.GetName())
```
Comment (sutaakar) [inline core/infrastructure/configmap.go:209]: Wouldn't Debug be enough here?

File: core/infrastructure/delta_processor.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	return &deltaProcessor{
+		context,
+	}
+}
+
+func (d *deltaProcessor) ProcessDelta(comparator compare.MapComparator, requestedResources map[reflect.Type][]resource.KubernetesResource, deployedResources map[reflect.Type][]resource.KubernetesResource) (isDeltaProcessed bool, err error) {
+	deltas := comparator.Compare(deployedResources, requestedResources)
+	for resourceType, delta := range deltas {
+		if !delta.HasChanges() {
+			d.Log.Info("No delta found", "resourceType", resourceType)
```
Comment (sutaakar) [inline core/infrastructure/delta_processor.go:45]:
```suggestion
			d.Log.Debug("No delta found", "resourceType", resourceType)
```

File: core/kogitoservice/deployer_resources.go [hunk 11]
```diff
@@ -234,41 +190,9 @@ func (s *serviceDeployer) getComparator() compare.MapComparator {
...
-		volumes = append(volumes, kogitoInfraInstance.GetStatus().GetVolumes()...)
-	}
-	return consolidateAppProperties, consolidateEnvProperties, volumes, nil
-}
-
 func (s *serviceDeployer) mountKogitoInfraVolumes(kogitoInfraVolumes []api.KogitoInfraVolumeInterface, deployment *appsv1.Deployment) {
-	appPropsVolumeHandler := NewAppPropsVolumeHandler()
-	framework.AddVolumeToDeployment(deployment, appPropsVolumeHandler.CreateAppPropVolumeMount(), appPropsVolumeHandler.CreateAppPropVolume(s.instance))
+	//appPropsVolumeHandler := NewAppPropsVolumeHandler()
+	//framework.AddVolumeToDeployment(deployment, appPropsVolumeHandler.CreateAppPropVolumeMount(), appPropsVolumeHandler.CreateAppPropVolume(s.instance))
```
Comment (sutaakar) [inline core/kogitoservice/deployer_resources.go:195]: Can this be deleted?

File: core/manager/supportingservice_manager.go [hunk 12]
```diff
@@ -108,7 +109,8 @@ func (k kogitoSupportingServiceManager) FetchKogitoSupportingServiceDeployment(n
 	return nil, nil
 }
 
-func (k kogitoSupportingServiceManager) getKogitoSupportingService(namespace string, serviceType api.ServiceType) (api.KogitoSupportingServiceInterface, error) {
+func (k kogitoSupportingServiceManager) FetchKogitoSupportingServiceForServiceType(namespace string, serviceType api.ServiceType) (api.KogitoSupportingServiceInterface, error) {
+	k.Log.Info("Fetching kogito Supporting Service", "serviceType", serviceType)
```
Comment (sutaakar) [inline core/manager/supportingservice_manager.go:113]:
```suggestion
	k.Log.Debug("Fetching kogito Supporting Service", "serviceType", serviceType)
```

**Comment by sutaakar** [comment]
Looks good to me, few comments above.

File: core/connector/protobuf_configmap_reconciler.go (hunk 8 above)
Comment (vaibhavjainwiz) [reply core/connector/protobuf_configmap_reconciler.go:101]: done

File: core/infrastructure/configmap.go (hunk 9 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/configmap.go:209]: changed to debug

File: core/infrastructure/delta_processor.go (hunk 10 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/delta_processor.go:45]: done

File: core/kogitoservice/deployer_resources.go (hunk 11 above)
Comment (vaibhavjainwiz) [reply core/kogitoservice/deployer_resources.go:195]: done

File: core/manager/supportingservice_manager.go (hunk 12 above)
Comment (vaibhavjainwiz) [reply core/manager/supportingservice_manager.go:113]: done

**Review by sutaakar — APPROVED** [review]

File: core/infrastructure/configmap.go (hunk 6 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/configmap.go:188]: changed name to `isConfigMapCreatedUsingLiteral`

File: api/v1beta1/kogitoruntime_types.go (hunk 1 above)
Comment (spolti) [reply api/v1beta1/kogitoruntime_types.go:122]: good catch, thanks.

**Review by spolti — APPROVED** [review]
@vaibhavjainwiz nice work :)

**Review by radtriste — APPROVED** [review]

**Review by ricardozanini — COMMENTED** [review]
Nice work! I'm just curious about this `core/connector` package. Why it's called "connector"? I get the service endpoint and URL use cases, but not the protobuf one.

Please, can you write a dedicated session in the Kogito Operator Documentation about configuration? I understand that now we have many possibilities, and we ought to pass the right directions to users, mainly with the product release. cc @hmanwani-rh @briandooley (this might be a critical doc issue).

**Comment by vaibhavjainwiz** [comment]
> dedicated session in the Kogito Operator Documentation about configuration?

I put those thing in connector which connect both runtime and supportingservice otherwise it will give cyclic dependency. For proto we have similiar situation because runtime expose these file which need to be mounted on data-index service.

Currently I am working on kogito-infra improvement which lot more options to user to configure runtime service. I will add `configuration` section in kie-doc with kogito-infra improvement.

**Comment by ricardozanini** [comment]
> I put those thing in connector which connect both runtime and supportingservice otherwise it will give cyclic dependency. For proto we have similiar situation because runtime expose these file which need to be mounted on data-index service.

No worries, I was just curious. Even though, in situations like this, you can consider creating a new package to avoid misinterpretation.

**Comment by vaibhavjainwiz** [comment]
> > I put those thing in connector which connect both runtime and supportingservice otherwise it will give cyclic dependency. For proto we have similiar situation because runtime expose these file which need to be mounted on data-index service.
> 
> No worries, I was just curious. Even though, in situations like this, you can consider creating a new package to avoid misinterpretation.

I had moved `protobuf` content to shared package.

**Pull request merged**